[wip]OTA-1010: extract included manifests with net-new capabilities

hongkailiu opened the pull request.

hongkailiu commented on pkg/cli/admin/release/extract_tools.go:
```diff
@@ -1167,6 +1168,100 @@ func copyAndReplace(errorOutput io.Writer, w io.Writer, r io.Reader, bufferSize
+	manifestInclusionConfiguration manifestInclusionConfiguration,
+	currentImplicitlyEnabled sets.Set[configv1.ClusterVersionCapability],
+) sets.Set[configv1.ClusterVersionCapability] {
+	// TODO Replace this with the function from lib-go
```
https://github.com/openshift/cluster-version-operator/pull/1133/files

wking commented on pkg/cli/admin/release/extract_tools.go:
```diff
@@ -1281,8 +1369,66 @@ func findClusterIncludeConfig(ctx context.Context, restConfig *rest.Config) (man
+			manifestInclusionConfiguration.Overrides,
+			true,
+		)
+		// update manifest is enabled, no need to check
```
hmm, but with `IncludeAllowUnknownCapabilities`, aren't the ones that fail the ones we can ignore?  I'd guess they meant "not even with all these new caps enabled will this manifest get included".  The ones we'll implicitly enable caps for are the ones that fail to get included if we apply the current caps, but which do get included if we allow new caps, and which match `SameResourceID`.

hongkailiu replied:
This function is refactored from this one in CVO: https://github.com/openshift/cluster-version-operator/blob/e90705bb5457b2b9d449b4377036c3de6617ebb6/pkg/payload/payload.go#L247-L252

My understanding is very weak on this function (although i have read it a couple of times already).

> hmm, but with IncludeAllowUnknownCapabilities, aren't the ones that fail the ones we can ignore?

No. Passed ones (`updateManErr == nil` in the next line) are ignored. The manifests that are included on the cluster wont generate new implicitly enabled capabilities.

hongkailiu commented:
/retest-required

petr-muller commented:
/cc

hongkailiu commented:
/retest-required

hongkailiu commented:
Some testing result from 0431ce493c3c4e5038f5dc919aafa30861aa3d34

Cluster-bot:

> launch 4.13.12 aws

```
$ make oc
$ ./oc adm release extract --included --credentials-requests --to credentials-requests quay.io/openshift-release-dev/ocp-release:4.14.0-rc.0-x86_64
I0304 13:36:44.744021   32452 extract_tools.go:1254] Those capabilities become implicitly enabled for the incoming release [ImageRegistry MachineAPI]
Extracted release payload from digest sha256:1d2cc38cbd94c532dc822ff793f46b23a93b76b400f7d92b13c1e1da042c88fe created at 2023-09-07T07:37:47Z

$ rg ImageRegistry credentials-requests
credentials-requests/0000_50_cluster-image-registry-operator_01-registry-credentials-request.yaml
6:    capability.openshift.io/name: ImageRegistry

$ ll credentials-requests
total 48
-rw-r--r--@ 1 hongkliu  staff   1.8K Mar  4 13:36 0000_30_machine-api-operator_00_credentials-request.yaml
-rw-r--r--@ 1 hongkliu  staff   738B Mar  4 13:36 0000_50_cloud-credential-operator_05-iam-ro-credentialsrequest.yaml
-rw-r--r--@ 1 hongkliu  staff   1.3K Mar  4 13:36 0000_50_cluster-image-registry-operator_01-registry-credentials-request.yaml
-rw-r--r--@ 1 hongkliu  staff   920B Mar  4 13:36 0000_50_cluster-ingress-operator_00-ingress-credentials-request.yaml
-rw-r--r--@ 1 hongkliu  staff   1.0K Mar  4 13:36 0000_50_cluster-network-operator_02-cncc-credentials.yaml
-rw-r--r--@ 1 hongkliu  staff   1.5K Mar  4 13:36 0000_50_cluster-storage-operator_03_credentials_request_aws.yaml

### without --included
$ rm -rf credentials-requests          
$ ./oc adm release extract --to credentials-requests quay.io/openshift-release-dev/ocp-release:4.14.0-rc.0-x86_64                                  
Extracted release payload from digest sha256:1d2cc38cbd94c532dc822ff793f46b23a93b76b400f7d92b13c1e1da042c88fe created at 2023-09-07T07:37:47Z

$ ll credentials-requests | wc -l
     682
```

hongkailiu commented:
/retest-required

hongkailiu commented:
/test okd-scos-e2e-aws-ovn

hongkailiu commented:
/test e2e-agnostic-ovn-cmd

hongkailiu commented on pkg/cli/admin/release/extract_tools.go:
```diff
@@ -1236,20 +1329,20 @@ func findClusterIncludeConfig(ctx context.Context, restConfig *rest.Config) (man
+			capSet = capabilitiesSpec.BaselineCapabilitySet
+		}
+		deepCopy := clusterVersion.Status.Capabilities.DeepCopy()
+		if capSet == configv1.ClusterVersionCapabilitySetCurrent {
```
`v4.y` could be grow too over time.
But it keeps the consistency with 
https://github.com/openshift/oc/blob/def90a65fcbad5d7e54add49975e9c3daa92a4f9/pkg/cli/admin/release/extract_tools.go#L1300-L1302

hongkailiu commented:
Rerun the test with def90a65fcbad5d7e54add49975e9c3daa92a4f9:

```
launch 4.13.12 aws
```

The cluster did not set `BASELINE_CAPABILITY_SET`. So it is the default value `vCurrent`.

```
$ oc get clusterversions.config.openshift.io version -o yaml | yq .spec
{
  "channel": "candidate-4.13",
  "clusterID": "01541c49-a5d6-4b02-ba6a-161a12c3ca79",
  "upstream": "https://api.integration.openshift.com/api/upgrades_info/graph"
}

$ ./oc adm release extract --included --credentials-requests --to credentials-requests quay.io/openshift-release-dev/ocp-release:4.14.0-rc.0-x86_64

I0314 09:39:57.809034   23241 extract_tools.go:1340] If the eventual cluster will not be the same minor version as this v4.2.0-alpha.0-2583-gdef90a6 'oc', the actual vCurrent capability set may differ.
I0314 09:39:57.809062   23241 extract_tools.go:1343] If the eventual cluster will not be the same minor version as this v4.2.0-alpha.0-2583-gdef90a6 'oc', the known capability sets may differ.
I0314 09:40:11.867239   23241 extract_tools.go:1253] Those capabilities become implicitly enabled for the incoming release []
Extracted release payload from digest sha256:1d2cc38cbd94c532dc822ff793f46b23a93b76b400f7d92b13c1e1da042c88fe created at 2023-09-07T07:37:47Z

$ rg ImageRegistry credentials-requests
credentials-requests/0000_50_cluster-image-registry-operator_01-registry-credentials-request.yaml
6:    capability.openshift.io/name: ImageRegistry

$ ll credentials-requests
total 48
-rw-r--r--@ 1 hongkliu  staff   1.8K Mar 14 09:40 0000_30_machine-api-operator_00_credentials-request.yaml
-rw-r--r--@ 1 hongkliu  staff   738B Mar 14 09:40 0000_50_cloud-credential-operator_05-iam-ro-credentialsrequest.yaml
-rw-r--r--@ 1 hongkliu  staff   1.3K Mar 14 09:40 0000_50_cluster-image-registry-operator_01-registry-credentials-request.yaml
-rw-r--r--@ 1 hongkliu  staff   920B Mar 14 09:40 0000_50_cluster-ingress-operator_00-ingress-credentials-request.yaml
-rw-r--r--@ 1 hongkliu  staff   1.0K Mar 14 09:40 0000_50_cluster-network-operator_02-cncc-credentials.yaml
-rw-r--r--@ 1 hongkliu  staff   1.5K Mar 14 09:40 0000_50_cluster-storage-operator_03_credentials_request_aws.yaml
```

Comparing with https://github.com/openshift/oc/pull/1958#issuecomment-2698571944, no caps became implicitly enabled as expected. Because they would be (explicitly) enabled with `BASELINE_CAPABILITY_SET=vCurrent`.

I wanted to try `BASELINE_CAPABILITY_SET=None` with

```
launch 4.13.12 aws,no-capabilities
```

But [cluster-bot is not so happy with the command](https://redhat-internal.slack.com/archives/CNHC2DK2M/p1741961332941969).

I expect to see some implicitly enabled caps there.

hongkailiu commented:
/test e2e-agnostic-ovn-cmd

hongkailiu commented:
/retest-required

hongkailiu commented on pkg/cli/admin/release/extract_tools.go:
```diff
@@ -1236,20 +1331,20 @@ func findClusterIncludeConfig(ctx context.Context, restConfig *rest.Config) (man
+			klog.Infof("If the eventual cluster will not be the same minor version as this %s 'oc', the actual %s capability set may differ.", reportedVersion, capSet)
+		}
+		deepCopy.EnabledCapabilities = append(deepCopy.EnabledCapabilities, configv1.ClusterVersionCapabilitySets[capSet]...)
+		klog.Infof("If the eventual cluster will not be the same minor version as this %s 'oc', the known capability sets may differ.", reportedVersion)
```
> Can we put this log line into an else branch, or something?

Done.

> We may also have a way to figure out the target version from the pullspec we're extracting

Left a TODO for this.

hongkailiu replied:
I cannot use the existing `ImageMetadataCallback`

https://github.com/openshift/oc/blob/88d00755325cddf1921c9d554265f83954f3103b/pkg/cli/image/extract/extract.go#L561-L563

because it is called too late.

petr-muller commented:
/cc

petr-muller commented on pkg/cli/admin/release/extract.go:
```diff
@@ -349,20 +354,42 @@ func (o *ExtractOptions) Run(ctx context.Context) error {
+				klog.V(2).Infof("Retrieved the version from image configuration in the image to extract: %s", v)
+				versionInImageConfig = v
+			} else {
+				klog.Info("Cannot retrieve the version from image configuration in the image to extract")
```
"Cannot" sounds bad so I'd expect this to be an error or warning but it is an info. Is that intentional?

hongkailiu replied:
It was intensional as `versionInImageConfig` is used only for the logging and I was not sure if it should be there for each release.
Now it is changed to the error level (and let us see how if anyone is going to complain of it).

This part of the code is moved to https://github.com/openshift/oc/pull/2050/files

petr-muller commented on pkg/cli/admin/release/extract.go:
```diff
@@ -349,20 +354,42 @@ func (o *ExtractOptions) Run(ctx context.Context) error {
+		if c := imageConfig.Config; c != nil {
+			if v, ok := c.Labels["io.openshift.release"]; ok {
+				klog.V(2).Infof("Retrieved the version from image configuration in the image to extract: %s", v)
+				versionInImageConfig = v
```
Can this callback be called multiple times, overwriting previous values in `versionInImageConfig``? If yes, can the callback be called in parallel, which would be a race?

hongkailiu replied:
If I understand it correctly, `oc adm release extract` has only one release to extract and one image means one image config.

From my test, it is called only once.

petr-muller replied:
Can we encode that assumption and blew up with a panic or Fatal if `versionInImageConfig` is already set?

hongkailiu replied:
The code has been moved to https://github.com/openshift/oc/pull/2050/files.
But I will keep this open because I did not address this comment and you might still think I should do it in this pull.

The concern is valid in general but it is unlikely to happen here.
The other callbacks of `extract.ExtractOptions` are not multi-thread safe either.

petr-muller commented on pkg/cli/admin/release/extract.go:
```diff
@@ -349,20 +354,42 @@ func (o *ExtractOptions) Run(ctx context.Context) error {
 
-	var manifestErrs []error
+	var needEnabledCapabilities bool
+	inclusionConfig := manifestInclusionConfiguration{}
```
```suggestion
	var inclusionConfig manifestInclusionConfiguration
```

for consistency with other zero value initializations above

petr-muller commented on pkg/cli/admin/release/extract.go:
```diff
@@ -349,20 +354,42 @@ func (o *ExtractOptions) Run(ctx context.Context) error {
-			inclusionConfig := manifestInclusionConfiguration{}
+			clientVersion, reportedVersion, err := version.ExtractVersion()
+			if err != nil {
+				return err
```
This is complicated code, can we wrap this in a `fmt.Errorf("... %w")` and add some context about what the code is trying to do

hongkailiu replied:
The relevant code is moved to https://github.com/openshift/oc/pull/2050

To be honest, I cannot really tell what `version.ExtractVersion()` is doing from its doc:

https://github.com/openshift/oc/blob/267009391e55919093984b63c22d67cb2de0c700/pkg/version/version.go#L79-L83

The code exists before this pull.
From the logging later, `reportedVersion` is used as the oc-cli's version.

petr-muller commented on pkg/cli/admin/release/extract.go:
```diff
@@ -349,20 +354,42 @@ func (o *ExtractOptions) Run(ctx context.Context) error {
 			if o.InstallConfig == "" {
-				inclusionConfig, err = findClusterIncludeConfig(ctx, o.RESTConfig)
+				needEnabledCapabilities = true
+				inclusionConfig, err = findClusterIncludeConfig(ctx, o.RESTConfig, reportedVersion, versionInImageConfig)
```
I do not understand how and why `versionInImageConfig` is used here (and on L392 too). That variable can be written by the `opts.ImageConfigCallback`, and until that callback is called, it has a string zero value: empty string. But I do not see any opportunity between here (when it is used) and L358 where that callback would be actually called, so I'd say the variable must be empty string here? Am I missing something?

hongkailiu replied:
This one is in the `TarEntryCallback` (when `versionInImageConfig` is used) while `versionInImageConfig` is set up in `ImageConfigCallback` which is called earlier than `TarEntryCallback`.

We could use the existing `ImageMetadataCallback` but it is called too late.
https://github.com/openshift/oc/pull/1958#discussion_r2007764478

petr-muller commented on pkg/cli/admin/release/extract.go:
```diff
@@ -377,11 +404,10 @@ func (o *ExtractOptions) Run(ctx context.Context) error {
 
-		tarEntryCallbacks = append(tarEntryCallbacks, func(hdr *tar.Header, _ extract.LayerInfo, r io.Reader) (bool, error) {
-			if hdr.Name == "image-references" && !o.CredentialsRequests {
+		manifestsCallbacks = append(manifestsCallbacks, func(name string, ms []manifest.Manifest, r io.Reader, enabled []configv1.ClusterVersionCapability) (cont bool, err error) {
```
It would be helpful for me if each append to `manifestCallback` had a brief comment that said what the callback is doing

hongkailiu replied:
`tarEntryCallbacks` is removed in https://github.com/openshift/oc/pull/2048/files

petr-muller commented on pkg/cli/admin/release/extract.go:
```diff
@@ -493,39 +515,72 @@ func (o *ExtractOptions) Run(ctx context.Context) error {
-			for i, callback := range tarEntryCallbacks {
-				if tarEntryCallbacksDone[i] {
+	manifestsCallback := func(name string, ms []manifest.Manifest, r io.Reader, enabled []configv1.ClusterVersionCapability) (bool, error) {
+		if len(manifestsCallbacks) > 0 {
```
again guard clause would make this more idiomatic

```
if len() == 0 {
  return true, nil
}
...
```

petr-muller reviewed (commented):
I only went through the `pkg/cli/admin/release/extract.go` diff and left some comments. The code is very hard to read, which is not your fault - it is mostly caused by constructing a series of very long anonymous callbacks that end up being called at whatever time later... Not sure what to do about it though. It would definitely help if this was a series of smaller PRs.

petr-muller commented:
It may be beyond the scope of what this PR attempts to do, but I have a feeling that the code could be made more readable if some of callbacks (that are currently lambdas using various option struct members and closures on the surrounding scope variables) were extracted into a dedicated, named and documented single-purpose types with methods that would be used as the callbacks, with a constructor that makes the callback inputs a specified interface.

hongkailiu commented:
> It may be beyond the scope of what this PR attempts to do, but I have a feeling that the code could be made more readable if some of callbacks (that are currently lambdas using various option struct members and closures on the surrounding scope variables) were extracted into a dedicated, named and documented single-purpose types with methods that would be used as the callbacks, with a constructor that makes the callback inputs a specified interface.

I will give this a try.
(Currently I have another series of pulls to merge. They are also opened long time ago and made some progress recently. I think I can close them quickly. I will come back to this one right after).

hongkailiu commented on pkg/cli/admin/release/extract.go:
```diff
@@ -349,20 +354,42 @@ func (o *ExtractOptions) Run(ctx context.Context) error {
+			klog.Info("Cannot retrieve the version because no image configuration is provided in the image to extract")
+			return
+		}
+		if c := imageConfig.Config; c != nil {
```
The relevant part are moved to https://github.com/openshift/oc/pull/2050
